### ISSUE TITLE
Missing working subscription operations for read-only models

### ISSUE THREAD

**Issue by lawmicha** [aws-amplify/amplify-cli#4182]
**Describe the bug**
Related to https://github.com/aws-amplify/amplify-cli/issues/2715
Opening this issue to provide a simple example of when I provide read permissions to all users, I do not have a way to subscribe to the data to get updates.

**Amplify CLI Version**
`4.18.0`

**To Reproduce**
1. Allow all users to be able to read the model
```graphql
type ModelReadOnly
  @model
  @auth(rules: [ { allow: owner, operations: [create, update, delete] } ])
{
  id: ID!
  description: String!
}
```
2. CreateMutation with owner=111
3. Query from owner=222, get back model instance with owner 111
4. There is no subscription operation which does not take in `owner` field to allow subscribing to onCreate, onUpdate, onDelete of this model instance containing owner 111 from owner 222's perspective

5. Allow all users to be able to read/update the model
```graphql
type ModelReadUpdate
  @model
  @auth(rules: [ { allow: owner, operations: [create, delete] } ])
{
  id: ID!
  description: String!
}
```
6. CreateMutation with owner=111
7. Query from owner=222, get back model instance with owner 111
8. onUpdate subscriptiono is now public (does not require owner field), owner 222 subscribes to onUpdate
8. owner=111 does update mutation on previously created model from step 6. owner 222 gets the mutation response.
9. onCreate and onDelete requires owner to be passed in, there is no way for owner 222 to get the models created and deleted by owner 111, but only updates.

**Expected behavior**
in the first model schema, when owner 222 can read the other owners model instance, there should be a subscription that allows to get mutation events for created/updated/deleted.
in the second model schema, when owner 222 can read and update the other owners model instance, there should be a way to get mutation events for created/deleted.

So there should be subscriptions which do not take in the owner field and return data for all model creations/updates/deletes

**Comment by lawmicha** [comment]
currently, the logic is that when there is an operation set on the auth directive, then the corresponding subscription will require owner field, for example
```
if operations contains 'create', inject ownerId in onCreate subscription
if operations contains 'update', inject ownerId in onUpdate subscription
if operations contains 'delete', inject ownerId in onDelete subscription
```
The expected behavior is whenever there is read permission by others, then subscriptions should be public
```
if operations does not contain 'read', do not inject ownerid in any of the subscriptions
if operations does contain 'read', then others do not have permission to read, then all subscriptions will require ownerId
```

If this understanding is correct, can we take a step further, and without having to inject ownerId at all, just like the mutations? the subscriptions already has access to the ownerId field through `$ctx.identity.username` and should return data for which i have permissions to read. the subscriptions should not expose data for data which i do not have permissions to read.

The odd case here is that if i restrict others to read my data, it doesn't matter if i do or do not restrict updates/deletes since others have no way to retrieve the data anyways.

**Comment by mdoesburg** [comment]
> So there should be subscriptions which do not take in the owner field and return data for all model creations/updates/deletes

I need this for my app too. I think my issue is related/similar: https://github.com/aws-amplify/amplify-cli/issues/3777

**Comment by brene** [comment]
@SwaySway - can you provide details if your PR is addressing this issue?

**Comment by SwaySway** [comment]
This PR addresses this issue by stating that if owner is allowed the following rule.
`@auth(rules: [ { allow: owner, operations: [create, update, delete] } ])`
`@auth` will not add authorization logic on the subscription operation unless read is added.

Closing this issue as the PR has been merged and released into version 4.21.1.

**Comment by lawmicha** [comment]
> This PR addresses this issue by stating that if owner is allowed the following rule.
> `@auth(rules: [ { allow: owner, operations: [create, update, delete] } ])`
> `@auth` will not add authorization logic on the subscription operation unless read is added.
> 
> Closing this issue as the PR has been merged and released into version 4.21.1.

@SwaySway  just trying to understand your PR changes, all libraries are currently built on the fact that as long as the operation is specified on auth rule then expect the corresponding subscription operation to require ownerId in the graphql input. This is not just behavior change but also API change, right?

Before
- if owner authRule contains create operation, onCreate subscription requires owner input
- if owner authRule contains update operation, onUpdate subscription requires owner input
- if owner authRule contains delete operation, onDelete subscription requires owner input

Now?
- if owner authRule contains read operation, onCreate|onUpdate|OnDelete requires owner input ?
- else do not add owner to input?

**Comment by lawmicha** [comment]
1. operations: [create, update, delete] => all subscription operations do not need owner field
2. operations: [create, update, delete, read] => all subscription operations require owner field
3. operation: [create, delete] => all subscription operations do not need owner field because ‘read’ is not there right?
4. operation: [update, delete] =>  same as 3
5. operation: [create, update] =>  same as 3
6. operation: [create, read] =>  all subscription operations require owner field
7. operation: [update, read] =>  same as 6
8. operation: [delete, read] =>  same as 6
9. operation: [read] =>  same as 6
10. operation no specified on owner auth rule =>  same as 2, operation = [create, update, read, delete] , right?
